Mathlib/CategoryTheory/Sites/EffectiveEpimorphic.lean

```diff
@@ -0,0 +1,417 @@
+/-
+Copyright (c) 2023 Adam Topaz. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Adam Topaz
+-/
+
+import Mathlib.CategoryTheory.Sites.Sieves
+import Mathlib.CategoryTheory.Limits.Shapes.KernelPair
+
+/-!
+
+# Effective epimorphisms
+
+We define the notion of effective epimorphic (pre)sieves, morphisms and family of morphisms
+and provide some API for relating the three notions.
+
+-/
+
+namespace CategoryTheory
+
+open Limits
+
+variable {C : Type _} [Category C]
+
+/-- A sieve is effective epimorphic if the associated cocone is a colimit cocone. -/
+def Sieve.EffectiveEpimorphic {X : C} (S : Sieve X) : Prop :=
+  Nonempty (IsColimit (S : Presieve X).cocone)
+
+/-- A presieve is effective epimorphic if the cocone assocaited to the sieve it generates
+is a colimit cocone. -/
+abbrev Presieve.EffectiveEpimorphic {X : C} (S : Presieve X) : Prop :=
+  (Sieve.generate S).EffectiveEpimorphic
+
+/--
+The sieve of morphisms which factor through a given morphism `f`.
+This is equal to `Sieve.generate (Presieve.singleton f)`, but has
+more convenient definitional properties.
+-/
+def Sieve.generateSingleton {X Y : C} (f : Y ⟶ X) : Sieve X where
+  arrows Z g := ∃ (e : Z ⟶ Y), e ≫ f = g
+  downward_closed := by
+    rintro W Z g ⟨e,rfl⟩ q
+    refine ⟨q ≫ e, by simp⟩
+
+lemma Sieve.generateSingleton_eq {X Y : C} (f : Y ⟶ X) :
+    Sieve.generate (Presieve.singleton f) = Sieve.generateSingleton f := by
+  ext Z ; intro g
+  constructor
+  · rintro ⟨W,i,p,⟨⟩,rfl⟩
+    exact ⟨i,rfl⟩
+  · rintro ⟨g,h⟩
+    exact ⟨Y,g,f,⟨⟩,h⟩
+
+/--
+This structure encodes the data required for a morphism to be an effective epimorphism.
```

Comment:
Can you add a bit of doc for those who are not familiar with descent arguments in Grothendieck sites? Also, `desc` is quite clear but `fac` for `factorization` is perhaps too short a name.

Reply:
The names `desc` `fac` and `uniq` are meant to reflect the components of `IsColimit`. This struct is essentially encoding the fact that `X` is a colimit of some large complicated diagram.